### ISSUE TITLE
Cask could share dependencies between development projects (and others)

### ISSUE THREAD

**Issue by phillord** [cask/cask#362]
Currently, cask downloads all the dependencies to the .cask directory in the
directory of the package. This is fine but takes up extra space, bandwidth and
is irritating when on a dodgy connection -- we have to download dependencies
that we may already have in another package.

It would be nice if cask downloaded instead to ~/.emacs.d/cask/, and then
linked in packages to the package local .cask; this is the approach than
leiningen uses for instance.

As a quick hack, the current cask link functionality could be used to achieve
this. Longer term hard links might be nicer.

**Comment by lunaryorn** [comment]
@phillord We rely on package.el to download and install packages—currently all we do is change `package-user-dir`.

If we were to share packages we would need to handle downloading and installation ourselves as package.el has no such concept.  While that's not impossible it bears the risk of deviating from how package.el works and thus break things, and it's a considerable task.  I don't have the time for that.

**Comment by lunaryorn** [comment]
Besides, sharing bears the risk of binary incompatibilities. If package A used a macro from B it could be installed and loaded against different versions of the macro and thus break in subtle and hard to debug ways.

These incompatibilities already occur within a single set of packages (I.e. Helm which is notorious for macro issues), imagine we'd have multiple sets of packages mixed together.

I don't think that this would work out well.

**Comment by phillord** [comment]
On Sun, August 7, 2016 9:41 am, Sebastian Wiesner wrote:
> @phillord We rely on package.el to download and install
> packages—currently all we do is change `package-user-dir`.
>
> If we were to share packages we would need to handle downloading and
> installation ourselves as package.el has no such concept.  While that's
> not impossible it bears the risk of deviating from how package.el works
> and thus break things, and it's a considerable task.  I don't have the
> time for that.



Can't we just reset the package.el install directory, install the package
where ever we want it (i.e. ~/.emacs.d/cask/elpa)? Then next time we run
cask, we use PROJ/.cask/elpa, and all is good?

**Comment by phillord** [comment]
On Sun, August 7, 2016 9:56 am, Sebastian Wiesner wrote:
> Besides, sharing bears the risk of binary incompatibilities. If package A
> used a macro from B it could be installed and loaded against different
> versions of the macro and thus break in subtle and hard to debug ways.
>
> These incompatibilities already occur within a single set of packages
> (I.e. Helm which is notorious for macro issues), imagine we'd have
> multiple sets of packages mixed together.
>
> I don't think that this would work out well.


I can see the risk if we install B, install A, then upgrade B. The byte
compiled A would be using the old version of the macro in B.

But then this risk is there at the moment. A must be recompiled (or
reinstalled). With multiple copies, it may need recompiling in several
packages.

**Comment by lunaryorn** [comment]
@phillord What if project B uses MELPA Stable but project A previously used MELPA?  How'd we make sure that project B uses dash.el from MELPA Stable even if it was previously installed from MELPA for project A?

----

For sure we already accept the risk of binary incompatibilities, but the risk is much lower for a single coherent set of dependencies.  If we were to install dependencies from various projects all into the same place the risk would increase, and it'd be even harder to debug as you'd have to take all dependencies of all local projects into account.

----

My vote's definitely against this feature.

**Comment by phillord** [comment]
On Sun, August 7, 2016 12:34 pm, Sebastian Wiesner wrote:
> @phillord What if project B uses MELPA Stable but project A previously
> used MELPA?  How'd we make sure that project B uses dash.el from MELPA
> Stable even if it was previously installed from MELPA for project A?

Version numbers, which are already included in elpa package directory names.

Installing Project B would require that we check to see the dependencies
that we should install, then install them to ~/.emacs.d, then symlink them
in.

Installing Project A, with melpa would download the dated version from
melpa, and symlink this instead.


> For sure we already accept the risk of binary incompatibilities, but the
> risk is much lower for a single coherent set of dependencies.  If we were
> to install dependencies from various projects all into the same place the
> risk would increase, and it'd be even harder to debug as you'd have to
> take all dependencies of all local projects into account.

I can't see this. Only the dependencies of the local project would be
relevant since only those would be linked in.

> My vote's definitely against this feature.

Are you objecting to the feature or the implementation (which doesn't
exist yet, of course)?

**Comment by lunaryorn** [comment]
@phillord I object to the implementation you proposed—namely getting away with installing to a shared location and linking packages over.

I'm sure that it'd mostly work well in practices, but it gives up almost all the guarantees about isolation that Cask currently provides, and if it breaks it breaks badly, with hard-to-debug errors.

----

> Version numbers, which are already included in elpa package directory names.

I'm not sure what you mean with that.  It's a pure coincidence that we can discern MELPA and MELPA Stable by their version numbers.  It doesn't work for other archives, i.e. MELPA Stable and GNU ELPA.   And we've got no guarantee that a package (e.g. company-0.9) is the same in MELPA Stable and GNU ELPA.  These archives have completely different build procedures.

If a project wants Company from MELPA Stable we've got to get it from MELPA Stable; but if we install to a shared location we can't, because package.el doesn't even track where a package came from, let alone let's us install the same version twice from different archives.

>Installing Project B would require that we check to see the dependencies
> that we should install, then install them to ~/.emacs.d, then symlink them
> in.

And what about the dependencies of the dependencies? We can control where we install the immediate dependencies of project B from, but we can't control how package.el internally resolves the dependencies of those dependencies.

If we installed to a shared location it could well happen that a transitive dependency D of the immediate dependency C of project B was already installed via another dependency of project A, but from a different archive.  package.el would use the already installed dependency D, and thus break the isolation between different Cask environments, because all of a sudden we could get transitive dependencies from archives that the `Cask` file doesn't even specify.

----

TL;DR: If we install packages to a shared location by simply pointing `package-user-dir` to that location and afterwards linking all required packages to the target environment we can't uphold and guarantee isolation between Cask environments of different projects anymore.

We must implement **dependency installation** ourselves to track which archive a package came from for transitive dependencies, and we must implement **dependency resolution** ourselves to make sure that we only resolve against packages from archives contained in the `Cask` file of the project we're installing for. 

package.el does **not** support that, it doesn't even the internal data structures to support that—as I said it doesn't track the source of an installed package, and it can't restrict resolution to a single archive only.

And beyond that we also have to ensure that every immediate and transitive dependency of a project is byte-compiled with only the dependencies of the target project available so that any potential compile-time feature test runs only against the allowed set of packages.

So much for the implementation.

----

As for the idea, I don't think that we can cache installed and compiled packages without giving up isolation between Cask environments.  So as far as I'm concerned that's a no-go.  A package manager must favour correctness over speed, imho.

But we can cache downloads, imho.  We could

- cache downloaded packages per archive,
- teach package.el to look at the cache before downloading a package,
- and install packages regularly into the project-specific Cask directory, just like we do now.

That'd ensure that each package gets build within the isolated environment with no interaction with packages outside of that environment, and thus maintain isolation between the environments.

But that's just partial caching, at best, and it's a non-trivial implementation I figure as you'd have to hook into package.el's internals to teach it about the cache.

If you managed to implement that I'd happily review a pull request, but I for my part won't work on it by myself.  It's too much work for too little gain in my opinion.

**Comment by rejeep** [comment]
I usually don't want to add/change behaviour unless it's something that many people ask for. If someone asks for something, I "ignore" it the first time. If the person (or someone else) asks for the same thing again, then it might be worth considering.

That said, I have never had any issues with Cask being too slow and I've never heard anyone else complain about it either. Given the risks, I don't think it's worth the change.

**Comment by phillord** [comment]
Sebastian Wiesner <notifications@github.com> writes:

> But we can cache downloads, imho.  We could
>
> - cache downloaded packages per archive,
> - teach package.el to look at the cache before downloading a package,
> - and install packages regularly into the project-specific Cask directory, just like we do now.


Yes, that's true, a cache would also fulfil the requirements. I will
think on it.

**Comment by phillord** [comment]
Johan Andersson <notifications@github.com> writes:

> I usually don't want to add/change behaviour unless it's something that many
> people ask for. If someone asks for something, I "ignore" it the first time.
> If the person (or someone else) asks for the same thing again, then it might
> be worth considering.
>
> That said, I have never had any issues with Cask being too slow and I've never
> heard anyone else complain about it either. Given the risks, I don't think
> it's worth the change.


No worries, we will part the idea, until (and if) others complain.

Phil

**Comment by Fuco1** [comment]
I think we can close this, what do you think @phillord ? Emacs packages are so small that this makes little impact, unlike say Haskell's stack which can easily consume gigabytes of storage.

**Comment by dickmao** [comment]
The consensus appears quite clear.  Sharing dependencies lies madness, especially as the whole point of cask is to isolate.
